React/src/Chat.js

```diff
@@ -1,31 +1,52 @@
-import React, { useEffect, useState } from 'react';
-import { Avatar, IconButton } from '@mui/material';
-import { SearchOutlined, AttachFile, MoreVert } from '@mui/icons-material/';
-import { InsertEmoticon, Mic } from '@mui/icons-material/';
-import { useParams } from 'react-router-dom';
-import { useStateValue } from './StateProvider';
-import firebase from 'firebase/compat/app';
-import { _fetch, getFileURL, dateToLocal } from './global';
-import './Chat.css';
+import React, { useEffect, useState } from "react";
+import { Avatar, IconButton } from "@mui/material";
+import { SearchOutlined, AttachFile, MoreVert } from "@mui/icons-material/";
+import { InsertEmoticon, Mic } from "@mui/icons-material/";
+import { useParams } from "react-router-dom";
+import { useStateValue } from "./StateProvider";
+import firebase from "firebase/compat/app";
+import { _fetch, getFileURL, dateToLocal } from "./global";
+import "./Chat.css";
+import { actionTypes } from "./reducer";
 
 function Chat() {
     const { roomId } = useParams();
     const [input, setInput] = useState("");
     const [roomInfo, setRoomInfo] = useState({});
-    const [messages, setMessages] = useState([]);
-    const [{ user }, dispatch] = useStateValue();
+    // const [messages, setMessages] = useState([]);
+    const [{ user, messages, rooms }, dispatch] = useStateValue();
+    // const [{}, ] = useMessageStateValue();
+    const gotoLastMessageRef = React.useRef(null);
 
+    React.useEffect(() => {
+        gotoLastMessageRef.current.scrollIntoView({ behavior: "smooth" });
+    });
     useEffect(() => {
         if (roomId) {
             _fetch(user, "/api/ChatRoom/GetInfo?id=" + roomId)
-                .then(response => response.json())
-                .then(roomInfo => setRoomInfo(roomInfo));
+                .then((response) => response.json())
+                .then((roomInfo) => setRoomInfo(roomInfo));
 
             _fetch(user, "/api/Message/GetMany?chatRoomId=" + roomId)
-                .then(response => response.json())
-                .then(messages => setMessages(messages));
+                .then((response) => response.json())
+                .then((message) => {
+                    // setMessages(message);
+                    dispatch({
+                        type: actionTypes.FETCH_MESSAGES,
+                        messages: message,
+                    });
+                });
         }
-    }, [roomId, user]);
+
+        _fetch(user, "/api/ChatRoom/GetAll")
+        .then((response) => response.json())
+        .then((_rooms) => {
+            dispatch({
+                type: actionTypes.SET_ROOMS,
+                rooms: _rooms,
+            });
+        });
```

Comment:
Given the "/api/ChatRoom/GetInfo?id=" a few lines above, this portion here is not needed. It is only needed upon successful message submission, which you did implement.

```suggestion

```

Reply:
I notice that the need for this code is to achieve what the only useEffect() inside Sidebar.js was meant to achieve. It is more appropriate to do the dispatch (of rooms) inside the useEffect() of Sidebar.js than inside the useEffect() of Chat.js.

Reply:
> Given the "/api/ChatRoom/GetInfo?id=" a few lines above, this portion here is not needed. It is only needed upon successful message submission, which you did implement.
> 
> ```suggestion
> 
> ```

I get your point. I fetched the rooms here initially so I can save in the store.
Notice in Sidebar.js that the rooms are being fetched from the store as you earlier wanted. Removing this lines will cause the sidebar to be  empty